src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs

```diff
@@ -250,25 +251,79 @@ public TagHelperContent EndTagHelperWritingScope()
         /// </summary>
-        /// <param name="copyableTextWriter">Contains the data to be written.</param>
-        public void Write(ITextWriterCopyable copyableTextWriter)
+        /// <param name="tagHelperExecutionContext">The execution context containing the output.</param>
+        public async Task WriteTagHelperAsync([NotNull] TagHelperExecutionContext tagHelperExecutionContext)
         {
-            WriteTo(Output, copyableTextWriter);
+            await WriteTagHelperToAsync(Output, tagHelperExecutionContext);
         }
 
         /// <summary>
-        /// Writes an <see cref="ITextWriterCopyable"/> to the <paramref name="writer"/>.
+        /// Wrutes the output of a specified <see cref="TagHelperExecutionContext"/> to the specified
+        /// <paramref name="writer"/>.
         /// </summary>
-        /// <param name="writer">The <see cref="TextWriter"/> to which the
-        /// <paramref name="copyableTextWriter"/> is written.</param>
-        /// <param name="copyableTextWriter">Contains the data to be written.</param>
-        public void WriteTo([NotNull] TextWriter writer, ITextWriterCopyable copyableTextWriter)
+        /// <param name="writer">The <see cref="TextWriter"/> instance to write to.</param>
+        /// <param name="tagHelperExecutionContext">The execution context containing the output.</param>
+        public async Task WriteTagHelperToAsync(
+            [NotNull] TextWriter writer, [NotNull] TagHelperExecutionContext tagHelperExecutionContext)
+        {
+            var tagHelperOutput = tagHelperExecutionContext.Output;
+            var isTagNameNullOrWhitespace = string.IsNullOrWhiteSpace(tagHelperOutput.TagName);
+
+            if (!isTagNameNullOrWhitespace)
+            {
+                writer.Write('<');
+                writer.Write(tagHelperOutput.TagName);
+
+                foreach (var attribute in tagHelperOutput.Attributes)
+                {
+                    var value = HtmlEncoder.HtmlEncode(attribute.Value);
+                    writer.Write(' ');
+                    writer.Write(attribute.Key);
+                    writer.Write("=\"");
+                    writer.Write(value);
+                    writer.Write('"');
+                }
+
+                if (tagHelperOutput.SelfClosing)
+                {
+                    writer.Write(" /");
+                }
+
+                writer.Write('>');
+            }
+
+            if (isTagNameNullOrWhitespace || !tagHelperOutput.SelfClosing)
+            {
+                WriteTagHelperContentTo(writer, tagHelperOutput.PreContent);
+                if (tagHelperOutput.IsContentModified)
+                {
+                    WriteTagHelperContentTo(writer, tagHelperOutput.Content);
+                }
+                else if (tagHelperExecutionContext.ChildContentRetrieved)
+                {
+                    WriteTagHelperContentTo(writer, await tagHelperExecutionContext.GetChildContentAsync());
```

Comment:
break this bit into two since it's now in a helper method and not code gen'd:
```C#
var childContent = await tagHelperExecutionContext.GetChildContentAsync();
WriteTagHelperContentTo(writer, childContent);
```